politeiad/backendv2/tstorebe/store/mysql/mysql.go

```diff
@@ -0,0 +1,304 @@
+// atomically.
+//
+// This function satisfies the store BlobKV interface.
+func (s *mysql) Del(keys []string) error {
+	log.Tracef("Del: %v", keys)
+
+	ctx, cancel := ctxWithTimeout()
+	defer cancel()
+
+	// Start transaction
+	opts := &sql.TxOptions{
+		Isolation: sql.LevelDefault,
+	}
+	tx, err := s.db.BeginTx(ctx, opts)
+	if err != nil {
+		return err
+	}
+
+	// Delete blobs
+	for _, v := range keys {
+		_, err = tx.ExecContext(ctx, "DELETE FROM kv WHERE k IN (?);", v)
+		if err != nil {
+			// Attempt to roll back the transaction
+			if err2 := tx.Rollback(); err2 != nil {
+				// We're in trouble!
+				e := fmt.Sprintf("del: %v, unable to rollback: %v", err, err2)
+				panic(e)
+			}
+			return err
+		}
+	}
+
+	// Commit transaction
+	err = tx.Commit()
+	if err != nil {
+		return fmt.Errorf("commit: %v", err)
+	}
+
+	log.Debugf("Deleted blobs (%v) from store", len(keys))
+
+	return nil
+}
+
+// Get returns blobs from the store for the provided keys. An entry will not
+// exist in the returned map if for any blobs that are not found. It is the
+// responsibility of the caller to ensure a blob was returned for all provided
+// keys.
+//
+// This function satisfies the store BlobKV interface.
+func (s *mysql) Get(keys []string) (map[string][]byte, error) {
+	log.Tracef("Get: %v", keys)
+
+	ctx, cancel := ctxWithTimeout()
+	defer cancel()
+
+	// Build query. A placeholder parameter (?) is required for each
+	// key being requested.
+	//
+	// Ex 3 keys: "SELECT k, v FROM kv WHERE k IN (?, ?, ?)"
+	sql := "SELECT k, v FROM kv WHERE k IN ("
+	for i := 0; i < len(keys); i++ {
+		sql += "?"
+		// Don't add a comma on the last one
+		if i < len(keys)-1 {
+			sql += ","
+		}
+	}
+	sql += ");"
+
+	log.Tracef("%v", sql)
+
+	// The keys must be converted to []interface{} for the query method
+	// to accept them.
+	args := make([]interface{}, len(keys))
+	for i, v := range keys {
+		args[i] = v
+	}
+
+	// Get blobs
+	rows, err := s.db.QueryContext(ctx, sql, args...)
+	if err != nil {
+		return nil, fmt.Errorf("query: %v", err)
+	}
+	defer rows.Close()
+
+	reply := make(map[string][]byte, len(keys))
+	for rows.Next() {
+		var k string
+		var v []byte
+		err = rows.Scan(&k, &v)
+		if err != nil {
+			return nil, fmt.Errorf("scan: %v", err)
+		}
+		reply[k] = v
+	}
+	err = rows.Err()
+	if err != nil {
+		return nil, fmt.Errorf("next: %v", err)
+	}
+
+	// Decrypt data blobs
+	for k, v := range reply {
+		encrypted := isEncrypted(v)
+		log.Tracef("Blob is encrypted: %v", encrypted)
+		if !encrypted {
+			continue
+		}
+		b, _, err := s.decrypt(v)
+		if err != nil {
+			return nil, fmt.Errorf("decrypt: %v", err)
+		}
+		reply[k] = b
+	}
+
+	return reply, nil
+}
+
+// Closes closes the blob store connection.
+func (s *mysql) Close() {
+	s.zeroKey()
+	s.db.Close()
+}
+
+func New(appDir, host, user, password, dbname, keyFile string) (*mysql, error) {
+	// Load encryption key
+	if keyFile == "" {
```

Comment:
We shouldn't store the key on the filesystem. Send in a password that is obtained from the environment so that it can be easily containerized. Use ARGON2 to derive the key from the password.

Reply:
Done.

https://github.com/decred/politeia/pull/1180/files#diff-c6f51806560f29a8abfb043579039b00de52b398f7831c31c647c9b4d748deb4R19-R76